### DISCUSSION TITLE
Lossless WebP?

### DISCUSSION THREAD

**Discussion by amoerie** [saucecontrol/PhotoSauce#88] [Q&A] [answered]
Hi! Thanks for your absolutely amazing library. I am investigating whether this is feasible to use in a radiology viewer application, so my investigation is restricted to using lossless formats.

For my investigation, I am creating benchmarks that try out the various image formats and analyze:
- Average encoding speed (this is not super important, can be done in a background task)
- Average decoding speed (this is very important, the user will be actively waiting for this to finish)
- Average compression ratio (this is very important, the user will be actively waiting to finish downloading the image)

I also have some unit tests that go through a single encode + decode and verify the pixel data is the same, which hopefully confirms that the encoding and decoding process is lossless.

You can follow my progress here: https://github.com/amoerie/lossless-codec-benchmarks 

For now, it looks like WebP and AVIF are not lossless yet, so I am presuming I'm configuring them wrong.
I also can't seem to get HEIF working, but I am also presuming an error on my part. Investigation there is ongoing.

This is my question: is it already possible to pass in WebP and AVIF encoder options so it can be set to lossless? I found a folder "Webp" under "NativeCodecs" but this only seems to implement a decoder, and is nowhere on NuGet yet. 

Furthermore, if you have advice on the best parameters to use per file format for my benchmarks, I would be very eager to hear them.

Thanks and have a nice day!

P.S. Feel free to (re)use the benchmarks and tests in any way if they can be of any use to you. 
P.P.S. The benchmarks also include LZ4 to compare a "naive" byte array encoding with the more intelligent image formats. Sorry if that's confusing.

**Comment by saucecontrol** [comment, marked as answer]
Howdy! I can fill in a few gaps for you.

First thing to know is that not all mime types known to MagicScaler have codec support presently.  `TrySetEncoderFormat` returns a bool to tell you whether an encoder for the requested format was found, so you would need to check that result.

AVIF is not supported at all yet, and although it does have a spec for lossless encoding, it's not good -- as in it's less efficient than PNG which is much much older.  If we add AVIF, it will probably be lossy only.

HEIF is supported only for decode, mostly because Apple is the only major user of that format, and we're only concerned with being able to process HEIC images from Apple devices and get them converted to something more widely used.

WebP support is in progress.  As you've seen, the decoder is done but not yet published to NuGet.  I'll put out a public package when the encoder is working too, and it will support lossless mode.  That's tracking under https://github.com/saucecontrol/PhotoSauce/issues/19

I'm curious about your use case, though.  Radiology images are typically 16-bit greyscale, and you'd need to preserve the full bit depth.  MagicScaler currently only supports 8-bit output, so even the lossless formats will reduce the bit depth during transcode.

The primary use for MagicScaler so far is processing for web output, where 16-bit doesn't make much sense.  There is a longer-term plan to allow higher bit-depths through the pipeline -- as HDR becomes more commonly supported, we'll need at least 10- or 12-bit output even for web.  I'd like to consider supporting medical imaging use cases as well but haven't put a lot of thought into it.  If you can share a bit more about your workflow, that would be helpful for planning purposes.

**Reply by amoerie** [reply]
> Howdy! I can fill in a few gaps for you.

Same-day response! What luxury. 🎉

> First thing to know is that not all mime types known to MagicScaler have codec support presently. `TrySetEncoderFormat` returns a bool to tell you whether an encoder for the requested format was found, so you would need to check that result.

That's okay, I'll pick one and then keep using that. It's not my intention to "discover" at runtime in production whether something is supported or not.

> AVIF is not supported at all yet, and although it does have a spec for lossless encoding, it's not good -- as in it's less efficient than PNG which is much much older. If we add AVIF, it will probably be lossy only.

I had already read something similar on the internet, but I wanted to see it with my own eyes. But okay, not an option then!

> HEIF is supported only for decode, mostly because Apple is the only major user of that format, and we're only concerned with being able to process HEIC images from Apple devices and get them converted to something more widely used.

Fair enough, same story as AVIF as far as I'm concerned then

> WebP support is in progress. As you've seen, the decoder is done but not yet published to NuGet. I'll put out a public package when the encoder is working too, and it will support lossless mode. That's tracking under #19

Great, thanks! Looking forward to trying that out, so far WebP seems one of the most promising candidates.

> I'm curious about your use case, though. Radiology images are typically 16-bit greyscale, and you'd need to preserve the full bit depth. MagicScaler currently only supports 8-bit output, so even the lossless formats will reduce the bit depth during transcode.

Don't worry. The viewer will automatically switch to raw 16 bit pixel data on demand if a certain function demands it. But by default, we use an 8-bit flow because it's simply much faster (we can send encoded data to the browser and let the browser decode it, highly optimized!). However, we also need to support backend endpoints that crop & resize to implement a sort of "tiling on demand" feature for high resolution images. In these scenarios, we need to decode on the backend, resize & crop and then re-encode again. Sending the raw bitmap to the frontend incurs a too high network penalty.

> The primary use for MagicScaler so far is processing for web output, where 16-bit doesn't make much sense. There is a longer-term plan to allow higher bit-depths through the pipeline -- as HDR becomes more commonly supported, we'll need at least 10- or 12-bit output even for web. I'd like to consider supporting medical imaging use cases as well but haven't put a lot of thought into it. If you can share a bit more about your workflow, that would be helpful for planning purposes.

See previous question. What we do today is send the 16 bit pixel data raw to the browser (LZ4 or Snappy zipped to save a bit of bandwidth) and then apply a window level algorithm that interpolates to an 8-bit output that we can draw on a regular HTML canvas. 
See http://what-when-how.com/embedded-image-processing-on-the-tms320c6000-dsp/windowlevel-image-processing-part-1/ for some more information, but mathematically it's just a linear or bilinear interpolation step.

The biggest thing we're missing today is a good 16 bit codec that can drastically reduce the file size we need to send over the wire, with a decoder that is either simple & fast enough to run in a javascript web worker or (preferably) a decoder built in to the browser. 

Because this is such a big gap in the web ecosystem, we store images in both 8 and 16 bit and deliver them in 8 bit by default for performance reasons. I believe many radiology web viewers do something similar.

**Reply by saucecontrol** [reply]
Ah, great.  Thanks for the extra info.  Does sound like lossless WebP would be a good fit for you, then.  That'll be done soon.

JPEG-XL is promising as well, but the current version of the encoder is very rough.  The libjxl team is very focused on the decoder side for now, as support in Chromium is dependent on that.  Once the decoder is stabilized, I expect the encoder will start to see improvement.

**Reply by amoerie** [reply]
Thank you too for making this great library!

Even though WebP is not ready yet, I'll accept it as an authoritative answer, if anyone else has the same question the answer is simple: have patience and wait for the WebP nuget package to appear. :-)

**Reply by saucecontrol** [reply]
Nuget package is up: https://www.nuget.org/packages/PhotoSauce.NativeCodecs.Libwebp/

I learned while implementing it that WebP doesn't actually support greyscale formats, so compression isn't as good as it could be, and there's no way to know on decode that the image isn't RGB.  Probably not a win over PNG, but it's there to try 🤷‍♂️

**Reply by amoerie** [reply]
Thanks! I'll play with it and give an update soon! 👍

**Reply by amoerie** [reply]
Hi again! 

Just to give a small update: the WebP implementation seems super promising! File sizes are 50% that of PNG, encoding is 5x faster and decoding is 2x faster on my sample data set.
We will probably schedule a proof of concept in our applications in the near future.

Thanks again for the great work!

**Reply by saucecontrol** [reply]
Very interesting, thanks for the update!
